nsq_api/service.py

```diff
@@ -0,0 +1,215 @@
+import sys
+import time
+import socket
+import json
+
+from tornado import gen, concurrent
+import tornado.ioloop
+import tornado.web
+import requests
+from nsq.reader import Reader, Message
+from basescript import BaseScript
+from deeputil import generate_random_string, AttrDict, keeprunning 
+from logagg_utils import NSQSender
+from logagg_utils import InvalidArgument, AuthenticationFailure
+from logagg_utils import start_daemon_thread, log_exception
+
+class NsqAPI(tornado.web.RequestHandler):
+
+    NSQ_MAX_IN_FLIGHT = 200
+    CHANNEL_NAME_LENGTH = 6
+    DELETE_CHANNEL_URL = 'http://{}/channel/delete?topic={}&channel={}'
+    executor = concurrent.futures.ThreadPoolExecutor(5)
+
+    def initialize(self, log):
+
+        self.log = log
+
+    def _parse_query(self, query_arguments):
+        '''
+        query_arguments: {'topic': ['Heartbeat'], 'nsqd_tcp_address': ['195.201.98.142:4150']}
+        '''
+
+        nsqd_tcp_addresses=query_arguments['nsqd_tcp_address']
+        topic = query_arguments['topic'][0]
+        if query_arguments.get('empty_lines') == ['yes']:
+            empty_lines = True
+        else:
+            empty_lines = False
+
+        return nsqd_tcp_addresses, topic, empty_lines
+
+    def _remove_channel(self, nsqd_host,
+            topic, channel):
+        '''
+        Delete a Nsq channel
+        '''
+
+        nsqd_http_port = '4151'
+        nsqd_http_address = nsqd_host + ':' + nsqd_http_port
+
+        requests.post(self.DELETE_CHANNEL_URL.format(nsqd_http_address,
+            topic, channel))
+
+    @gen.coroutine
+    def get(self):
+        '''
+        Sample uri: /tail?nsqd_tcp_address=localhost:4150&topic=Heartbeat&empty_lines='yes/no'
+        '''
+        loop = tornado.ioloop.IOLoop.current()
+
+        nsqd_tcp_addresses, topic, empty_lines = self._parse_query(self.request.query_arguments)
```

Comment:
@supriyopaul This one will throw an error if someone making a request with less than three parameters please ensure that we have all the parameters that are required before doing something like this.

Reply:
And if all the query parameters are not there, what should I do?

Reply:
return back saying required parameters are missing

Reply:
But the user is expecting a straming response, It will cause errors at client side then.

Reply:
If user doesn't pass the parameters needed how will you stream the data

Reply:
Streaming does not allow to put returns in the function:
    `SyntaxError: 'return' with argument inside generator`
Plus client gets an internal server error, so it doesn't fail silently.

Reply:
And if someone uses the url wrong why is it the API's responsibility to correct it? I agree a proper error message but it'l be too much of a complexity for such a small problem.